README.md
MetaInstance
========
[![Gem Version](http://img.shields.io/gem/v/meta_instance.svg?style=flat-square)](http://badge.fury.io/rb/meta_instance)
[![Build Status](http://img.shields.io/travis/NullVoxPopuli/meta-instance.svg?style=flat-square)](https://travis-ci.org/NullVoxPopuli/meta-instance)
[![Code Climate](http://img.shields.io/codeclimate/github/NullVoxPopuli/meta-instance.svg?style=flat-square)](https://codeclimate.com/github/NullVoxPopuli/meta-instance)
[![Test Coverage](http://img.shields.io/codeclimate/coverage/github/NullVoxPopuli/meta-instance.svg?style=flat-square)](https://codeclimate.com/github/NullVoxPopuli/meta-instance)
[![Dependency Status](http://img.shields.io/gemnasium/NullVoxPopuli/MetaInstance.svg?style=flat-square)](https://gemnasium.com/NullVoxPopuli/MetaInstance)



## What Is It?

MetaInstance is a *convenient* and *safe* API for accessing and manipulating
an object's state.

## How Does It Work

MetaInstance adds a method to all objects called `#instance`. It returns
an `Instance` delegator that provides the full interface to the
object's state.

Of course, without implementing this in C, directly in the Ruby source,
we  are left to depend on the current provisions Ruby has for accessing
the state of an object. So there are some limitations here. However,
we implement the Ruby code in such a way as to minimize the downsides
by caching all the method definitions the Instance class will utilize.

## Usage

Let's use a very simple example class with which to demonstrate usage.

```ruby
    class Song
      attr_accessor :title
      attr_accessor :artist
      attr_accessor :year

      def initialize(title, artist, year)
        @title  = title
        @artist = artist
        @year   = year
      end
    end
```

Now we can create an instance of Song and work with it's state.

```ruby
    song = Song.new("Paranoid", "Black Sabbath", 1970)

    song.instance.variables
    # => [:@title, :@artist, :@year]

    song.instance.get(:title)
    # => "Paranoid"

    song.instance[:artist]
    # => "Black Sabbath"

    song.instance.to_h
    # => {:title => "Paranoid", :artist => "Black Sabbath", :year => 1970}
```

For a more complete set of usage examples see the [QED](http://rubyworks.github.com/instance/qed.html) documentation.


## Copyrights

Copyright &copy; 2014 [Rubyworks](http://rubyworks.github.io)

BSD-2-Clause License

See [LICENSE.txt](LICENSE.txt) file for license details.


HISTORY.md
# RELEASE HISTORY

## 1.1.0 | 2015-05-27

This release prepares for gemification.

Changes:

* Add RSpec Tests
* Prep for gem publishing
* Merge code with NullVoxPopuli/meta_instance
* New Architecture
* Rename to MetaInstance, replacing existing meta_instance gem


## 0.2.0 | 2014-02-02

This release adds an extra set of methods for modules and classes.

Changes:

* Add `method_definition` if object is a module or class.
* Add `method_definitions` if object is a module or class.


## 0.1.0 | 2014-02-01

This is the initial release of Instance, a class spun-off from
Ruby Facets.

Changes:

* Happy Release Day!


spec/spec_helper.rb
require "rubygems"
require "bundler/setup"

require "pry-byebug" # binding.pry to debug!

# Coverage
require "codeclimate-test-reporter"
ENV['CODECLIMATE_REPO_TOKEN'] = "313a19165c7a59866f44bbcfedabd28c6185dbe5aae0ee4314cc937ca4cd443b"
CodeClimate::TestReporter.start

# This Gem
require "meta_instance"


# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# Require this file using `require "spec_helper"` to ensure that it is only
# loaded once.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  config.treat_symbols_as_metadata_keys_with_true_values = true
  config.run_all_when_everything_filtered = true
  config.filter_run :focus

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = 'random'
end


spec/instance_spec.rb
require 'spec_helper'

class Song
  attr_accessor :title
  attr_accessor :artist
  attr_accessor :year

  def initialize(title, artist, year)
    @title  = title
    @artist = artist
    @year   = year
  end
end

describe MetaInstance do

  let(:song){ Song.new("Paranoid", "Black Sabbath", 1970) }

  describe '#variables' do

    it 'returns instance variables as an array of symbols' do
      actual = song.instance.variables
      expected = [:@title, :@artist, :@year]
      expect(actual).to eq expected
    end

  end

  describe '#get' do
    it 'gets the instance variable' do
      actual = song.instance.get(:title)
      expect(actual).to eq 'Paranoid'
    end
  end

  describe '#set' do
    it 'sets the instance varibale' do
      expected = song.instance.get(:title) + '!!!'
      song.instance.set(:title, expected)
      actual = song.instance.get(:title)

      expect(actual).to eq expected
    end
  end

  describe '#[]' do
    it 'retrieves an instance variable' do
      actual = song.instance[:title]
      expect(actual).to eq 'Paranoid'
    end
  end

  describe '#to_h' do
    it 'converts the list of instance variables to a hash' do
      actual = song.instance.to_h
      expected = {:title => "Paranoid", :artist => "Black Sabbath", :year => 1970}

      expect(actual).to eq expected
    end

    it 'converts the list of instance variables to a hash with @ in the keys'  do
      actual = song.instance.to_h(true)
      expected = {:@title => "Paranoid", :@artist => "Black Sabbath", :@year => 1970}

      expect(actual).to eq expected
    end
  end

  describe '#size' do
    it 'counts the number of instance variables' do
      actual = song.instance.size
      expect(actual).to eq 3
    end
  end

  describe '#<<' do
    it 'sets an instance variable' do
      song.instance << [:something, :else]
      actual = song.instance.get(:something)

      expect(actual).to eq :else
    end
  end

  describe '#remove' do
    it 'removes an instance variable' do
      song.instance.remove(:title)
      actual = song.instance.get(:title)

      expect(actual).to eq nil
    end
  end

  describe '#update' do
    it 'updates an instance variable' do
      song.instance.update(title: 'new title', other: 'var')
      actual = song.instance.get(:title)
      actual2 = song.instance.get(:other)

      expect(actual).to eq 'new title'
      expect(actual2).to eq 'var'
    end
  end

  describe '#names' do
    it 'returns the list of instance variables' do
      actual = song.instance.keys

      expect(actual).to eq [:title, :artist, :year]
    end
  end

  describe '#values' do
    it 'retrurns a list of values of the instance variables' do
      actual = song.instance.values
      expected = ["Paranoid", "Black Sabbath", 1970]

      expect(actual).to eq expected
    end
  end


end


spec/meta_instance/instance_method_define_spec.rb
class Foo
  include MetaInstance::InstanceMethodDefine

  def bar; "bar"; end
end


require 'spec_helper'

describe MetaInstance::InstanceMethodDefine do
  let(:original){"bar"}
  let(:f){ Foo.new }

  it "returns the original value" do
    expect(f.bar).to eq original
  end

  context "define_method" do

    context "without args" do

      before(:each) do
        f.define_method(:bar) do
          "foo"
        end
      end

      it "redefines the method" do
        expect(f.bar).to eq "foo"
      end

      it "returns the redefined value on another instance" do
        expect(Foo.new.bar).to_not eq "foo"
      end

    end

    context "with args" do

      before(:each) do
        f.define_method(:bar) do |number|
          number * 2
        end

        it "takes an argument" do
          expect(f.bar(3)).to eq 6
        end
      end
    end
  end

  context "backup_method" do
    before(:each) do
      f.backup_method(:bar)
    end

    it "has the original method accessible" do
      name = "#{MetaInstance::InstanceMethodDefine::METHOD_BACKUP_KEY}bar"
      expect(f.send(name)).to eq original
    end
  end

  context "restore_method" do
    before(:each) do
      f.backup_method(:bar)
      f.define_method(:bar) { "different" }
      f.restore_method(:bar)
    end

    it "returns the original" do
      expect(f.bar).to eq original
    end

    it "doesn't respond_to the backup method" do
      expect(f).to_not respond_to "#{MetaInstance::InstanceMethodDefine::METHOD_BACKUP_KEY}bar"
    end

    it "doesn't remove a method that isn't there" do
      expect(f).to_not respond_to(:foobar)
      expect(f.restore_method(:foobar)).to be_nil
    end
  end

  context "instance_override" do
    before(:each) do
      f.instance_override(:bar) do
        4
      end
    end

    it "returns the new value" do
      expect(f.bar).to eq 4
    end

    it "has the original method accessible" do
      name = "#{MetaInstance::InstanceMethodDefine::METHOD_BACKUP_KEY}bar"
      expect(f.send(name)).to eq original
    end

    it "returns the redefined value on another instance" do
      expect(Foo.new.bar).to_not eq "foo"
    end

    context "second call on instance_override" do
      before(:each) do
        f.instance_override(:bar) do
          0
        end
      end

      it "does not re-backup the method" do
        name = "#{MetaInstance::InstanceMethodDefine::METHOD_BACKUP_KEY}bar"
        expect(f.send(name)).to eq original
      end

      it "returns the new value" do
        expect(f.bar).to eq 0
      end

    end

  end

end


lib/meta_instance.rb
# Instance class is a delgator for any object which provides an elegant
# and protected interface to an object's state, i.e. its *instance*.
# Elgence is achieved by providing a single interface, the `instance`
# method. Protection is made possible by caching all the built-in
# Ruby methods used to interface with an object's internal state.
# This way they can not be overriden by some errant code or third
# party library.
#
# Examples
#
#   class Friend
#     attr_accessor :name, :age, :phone
#     def initialize(name, age, phone)
#       @name, @age, @phone = name, age, phone
#     end
#   end
#
#   f1 = Friend.new("John", 30, "555-1212")
#   f1.instance.get(:name)  #=> "John"
#   f1.instance.update(:name=>'Jerry')
#   f1.instance.get(:name)  #=> "Jerry"
#

require 'active_support'

require 'meta_instance/version'
require 'meta_instance/freeze_method'
require 'meta_instance/module_extensions'
require 'meta_instance/instance_method_define'
require 'meta_instance/proxy'


class BasicObject
  # Returns an instance of Instance for `self`, which allows convenient
  # access to an object's internals.
  def instance
    ::MetaInstance::Proxy.instance(self)
  end
end


lib/meta_instance/proxy.rb
# require 'forwardable'
require 'active_support/core_ext'

module MetaInstance
  class Proxy
    # extend Forwardable
    include Enumerable
    include FreezeMethod
    include ModuleExtensions
    include InstanceMethodDefine

    freeze_method :object_id
    freeze_method :class
    freeze_method :instance_of?
    freeze_method :method
    freeze_method :methods
    freeze_method :public_methods
    freeze_method :protected_methods
    freeze_method :private_methods
    freeze_method :instance_eval
    freeze_method :instance_exec
    freeze_method :instance_variables
    freeze_method :instance_variable_get
    freeze_method :instance_variable_set
    freeze_method :instance_variable_defined?
    freeze_method :remove_instance_variable
    freeze_method :send
    freeze_method :is_a?
    freeze_method :kind_of?

    delegate :send, :class, :kind_of, :is_a?, :instance_variables, to: :@delegate



    # Instance cache acts as a global cache for instances of Instance.
    @cache = {}

    # Instance is multiton. Use this method instead of #new to get a
    # cached instance.
    def self.instance(delegate)
      @cache[delegate] ||= MetaInstance::Proxy.new(delegate)
    end

    # Initialize new Instance instance. If the delegate is a type of
    # Module or Class then the instance will be extended with the
    # {ModuleExtensions} mixin.
    #
    def initialize(delegate)
      @delegate = delegate
      extend ModuleExtensions if Module === delegate
    end

    # The delegated object.
    def delegate
      @delegate
    end

    # Iterate over instance variables.
    def each
      variables.each do |name|
        yield(name[1..-1].to_sym, get(name))
      end
    end

    # Number of instance variables.
    def size
      variables.size
    end

    # Get instance variables with values as a hash.
    #
    # Examples
    #
    #   class X
    #     def initialize(a,b)
    #       @a, @b = a, b
    #     end
    #   end
    #
    #   x = X.new(1,2)
    #
    #   x.instance.to_h  #=> { :a=>1, :b=>2 }
    #
    # Returns [Hash].
    def to_h(at=false)
      h = {}
      if at
        variables.each do |name|
          h[name] = get(name)
        end
      else
        each do |key, value|
          h[key] = value
        end
      end
      h
    end

    # Get instance variable's value. Will return `nil` if the
    # variable does not exist.
    #
    # Returns the value of the instance variable.
    def get(name)
      name = atize(name)
      delegate.instance_variable_get(name)
    end
    alias :[] :get

    # Set instance variable.
    #
    # Returns the set value.
    def set(name, value)
      name = atize(name)
      delegate.instance_variable_set(name, value)
    end
    alias :[]= :set

    # Set an instance variable given a name and a value in an array pair.
    #
    # Example
    #
    #   f = Friend.new
    #   f.instance << [:name, "John"]
    #   f.name #=> "John"
    #
    # Returns the set value.
    def <<(pair)
      name, value = *pair
      name = atize(name)
      set(name, value)
    end

    # Remove instance variable.
    def remove(name)
      name = atize(name)
      delegate.remove_instance_variable(name)
    end

    # Set instance variables given a +hash+.
    #
    #   instance.update('@a'=>1, '@b'=>2)
    #   @a   #=> 1
    #   @b   #=> 2
    #
    # Also, +@+ sign is not neccessary.
    #
    #   instance.update(:a=>1, :b=>2)
    #   @a   #=> 1
    #   @b   #=> 2
    #
    # Returns nothing.
    def update(hash)
      hash.each do |pair|
        self << pair
      end
    end

    # A hold-over from the the old #instance_assign method.
    alias_method :assign, :update

    # Same as #instance_variables.
    def variables
      delegate.instance_variables
    end



    # Instance vairable names as symbols.
    #
    # Returns [Array<Symbols>].
    def keys
      variables.collect do |name|
        name[1..-1].to_sym
      end
    end

    alias_method :names, :keys

    # Instance variable values.
    #
    # Returns [Array<Object>].
    def values
      variables.collect do |name|
        get(name)
      end
    end

    # Instance evaluation.
    def eval(*a,&b)
      delegate.instance_eval(*a, &b)
    end

    # Instance execution.
    def exec(*a,&b)
      delegate.instance_exec(*a, &b)
    end

    # Get method. Usage of this might seem strange because Ruby's own
    # `instance_method` method is a misnomer. It should be something
    # like `definition` or `method_definition`. In Ruby the acutal
    # "instance" method is accessed via the unadorned `method` method.
    #
    # Returns [Method].
    def method(name)
      bind_call(:method, name)
    end

    # Returns list of method names.
    #
    # Returns [Array<Symbol>].
    def methods(*selection)
      list = []

      if selection.empty?
        list.concat @delegate.methods
      end

      selection.each do |s|
        case s
        when :public, :all
          list.concat @delegate.public_methods
        when :protected, :all
          list.concat @delegate.protected_methods
        when :private, :all
          list.concat @bind_call.private_methods
        end
      end

      return list
    end

    # Is the object an instance of a given class?
    #
    # Returns [Boolean]
    def of?(a_class)
      delegate.instance_of?(a_class)
    end


    # Is an instaance variable defined?
    #
    # Returns [Boolean]
    def variable_defined?(name)
      name = atize(name)
      delegate.instance_variable_defined?(name)
    end

    # Get object's instance id.
    #
    # Returns [Integer]
    def id
      delegate.object_id
    end

    # Fallback to get the real class of the Instance delegate itself.
    alias :object_class :class

    private

    def atize(name)
      name.to_s !~ /^@/ ? "@#{name}" : name
    end

    # TODO: Are there any method we need specific to a Class vs a Module?
    #module ClassExtensions
    #
    #end

  end
end


lib/meta_instance/module_extensions.rb
require_relative "freeze_method"

##
# ModuleExtensions provides some additional methods for Module and Class
# objects.
#
# TODO: Are there any other module/class methods that need to be provided?
#
module MetaInstance::ModuleExtensions
  extend ActiveSupport::Concern

  include MetaInstance::FreezeMethod
  # Store Object methods so they cannot be overriden by the delegate class.

  included do
    freeze_method :instance_method
    freeze_method :instance_methods
    freeze_method :public_instance_methods
    freeze_method :protected_instance_methods
    freeze_method :private_instance_methods
  end

  # List of method definitions in a module or class.
  #
  # selection - Any of `:public`, `:protected` or `:private` which
  #             is used to select specific subsets of methods.
  #
  # Returns [Array<Symbol>]
  def method_definitions(*selection)
    list = []

    if selection.empty?
      list.concat bind_call(:instance_methods)
    end

    selection.each do |s|
      case s
      when :public, :all
        list.concat bind_call(:public_instance_methods)
      when :protected, :all
        list.concat bind_call(:protected_instance_methods)
      when :private, :all
        list.concat bind_call(:private_instance_methods)
      end
    end

    return list
  end

  # Shorter alias for #method_definitions.
  alias :definitions :method_definitions

  # Get a first-class method definition object.
  #
  # Returns an unbound method object. [UnboundMethod]
  def method_definition(name)
    bind_call(:instance_method, name)
  end

  alias :definition :method_definition
end


lib/meta_instance/instance_method_define.rb
# This is based on a few things from
# http://reference.jumpingmonkey.org/programming_languages/ruby/ruby-metaprogramming.html
#
# allows the adding of methods to instances,
# but not the entire set of instances for a
# particular class
module MetaInstance::InstanceMethodDefine
  extend ActiveSupport::Concern

  # when a method is stubbed with snapshot data, we stare the
  # original method prixefixed with this:
  METHOD_BACKUP_KEY = "_mata_instance_backup_current_"

  # backs up and overrides a method.
  # but don't override if we already have overridden this method
  def instance_override(name, &block)
    unless respond_to?("#{METHOD_BACKUP_KEY}#{name}")
      backup_method(name)
    end
    define_method(name, &block)
  end

  # Adds methods to a singletonclass
  # define_singleton_method(name, &block) is the same as doing
  #
  # meta_eval {
  #   define_method(name, &block)
  # }
  def define_method(name, &block)
    define_singleton_method(name, &block)
  end

  # backs up a method in case we want to restore it later
  def backup_method(name)
    meta_eval {
      alias_method "#{METHOD_BACKUP_KEY}#{name}", name
    }
  end

  # the original method becomes reaccessible
  def restore_method(name)
    if respond_to?("#{METHOD_BACKUP_KEY}#{name}")
      meta_eval {
        alias_method name, "#{METHOD_BACKUP_KEY}#{name}"
        remove_method "#{METHOD_BACKUP_KEY}#{name}"
      }
    end
  end

  private

  # evals a block inside of a singleton class, aka
  #
  # class << self
  #   self
  # end
  def meta_eval(&block)
    singleton_class.instance_eval(&block)
  end

end


lib/instance.rb
# Instance class is a delgator for any object which provides an elegant
# and protected interface to an object's state, i.e. its *instance*.
# Elgence is achieved by providing a single interface, the `instance`
# method. Protection is made possible by caching all the built-in
# Ruby methods used to interface with an object's internal state.
# This way they can not be overriden by some errant code or third
# party library.
#
# Examples
#
#   class Friend
#     attr_accessor :name, :age, :phone
#     def initialize(name, age, phone)
#       @name, @age, @phone = name, age, phone
#     end
#   end
#
#   f1 = Friend.new("John", 30, "555-1212")
#   f1.instance.get(:name)  #=> "John"
#   f1.instance.update(:name=>'Jerry')
#   f1.instance.get(:name)  #=> "Jerry"
#
class Instance
  include Enumerable

  # Store Object methods so they cannot be overriden by the delegate class.
  METHODS = {}

  def self.freeze_method(name)
    METHODS[name.to_sym] = Object.instance_method(name)
  end

  freeze_method :object_id
  freeze_method :class
  freeze_method :instance_of?
  freeze_method :method
  freeze_method :methods
  freeze_method :public_methods
  freeze_method :protected_methods
  freeze_method :private_methods
  freeze_method :instance_eval
  freeze_method :instance_exec
  freeze_method :instance_variables
  freeze_method :instance_variable_get
  freeze_method :instance_variable_set
  freeze_method :instance_variable_defined?
  freeze_method :remove_instance_variable
  freeze_method :send
  freeze_method :is_a?
  freeze_method :kind_of?

  # Instance cache acts as a global cache for instances of Instance.
  @cache = {}

  # Instance is multiton. Use this method instead of #new to get a
  # cached instance.
  def self.instance(delegate)
    @cache[delegate] ||= Instance.new(delegate)
  end

  # Initialize new Instance instance. If the delegate is a type of 
  # Module or Class then the instance will be extended with the
  # {ModuleExtensions} mixin.
  # 
  def initialize(delegate)
    @delegate = delegate
    extend ModuleExtensions if Module === delegate
  end

  # The delegated object.
  def delegate
    @delegate
  end

  # Iterate over instance variables.
  def each
    variables.each do |name|
      yield(name[1..-1].to_sym, get(name))
    end
  end

  # Number of instance variables.
  def size
    variables.size
  end

  # Get instance variables with values as a hash.
  #
  # Examples
  #
  #   class X
  #     def initialize(a,b)
  #       @a, @b = a, b
  #     end
  #   end
  #
  #   x = X.new(1,2)
  #
  #   x.instance.to_h  #=> { :a=>1, :b=>2 }
  #
  # Returns [Hash].
  def to_h(at=false)
    h = {}
    if at
      variables.each do |name|
        h[name] = get(name)
      end
    else
      each do |key, value|
        h[key] = value
      end
    end
    h
  end

  # Get instance variable's value. Will return `nil` if the
  # variable does not exist.
  #
  # Returns the value of the instance variable.
  def get(name)
    name = atize(name)
    bind_call(:instance_variable_get, name)
  end
  alias :[] :get

  # Set instance variable.
  #
  # Returns the set value.
  def set(name, value)
    name = atize(name)
    bind_call(:instance_variable_set, name, value)
  end
  alias :[]= :set

  # Set an instance variable given a name and a value in an array pair.
  #
  # Example
  #
  #   f = Friend.new
  #   f.instance << [:name, "John"]
  #   f.name #=> "John"
  #
  # Returns the set value.
  def <<(pair)
    name, value = *pair
    name = atize(name)
    set(name, value)
  end

  # Remove instance variable.
  def remove(name)
    name = atize(name)
    bind_call(:remove_instance_variable, name)
  end

  # Set instance variables given a +hash+.
  #
  #   instance.update('@a'=>1, '@b'=>2)
  #   @a   #=> 1
  #   @b   #=> 2
  #
  # Also, +@+ sign is not neccessary.
  #
  #   instance.update(:a=>1, :b=>2)
  #   @a   #=> 1
  #   @b   #=> 2
  #
  # Returns nothing.
  def update(hash)
    hash.each do |pair|
      self << pair
    end
  end

  # A hold-over from the the old #instance_assign method.
  alias_method :assign, :update

  # Same as #instance_variables.
  def variables
    bind_call(:instance_variables)
  end

  # Instance vairable names as symbols.
  #
  # Returns [Array<Symbols>].
  def keys
    variables.collect do |name|
      name[1..-1].to_sym
    end
  end

  # Instance variable names as strings.
  #
  # Returns [Array<String>].
  def names
    variables.collect do |name|
      name[1..-1]
    end
  end

  # Instance variable values.
  #
  # Returns [Array<Object>].
  def values
    variables.collect do |name|
      get(name)
    end
  end

  # Instance evaluation.
  def eval(*a,&b)
    bind_call(:instance_eval, *a, &b)
  end

  # Instance execution.
  def exec(*a,&b)
    bind_call(:instance_exec, *a, &b)
  end

  # Get method. Usage of this might seem strange because Ruby's own
  # `instance_method` method is a misnomer. It should be something
  # like `definition` or `method_definition`. In Ruby the acutal
  # "instance" method is accessed via the unadorned `method` method.
  #
  # Returns [Method].
  def method(name)
    bind_call(:method, name)
  end

  # Returns list of method names.
  #
  # Returns [Array<Symbol>].
  def methods(*selection)
    list = []

    if selection.empty?
      list.concat bind_call(:methods)
    end

    selection.each do |s|
      case s
      when :public, :all
        list.concat bind_call(:public_methods)
      when :protected, :all
        list.concat bind_call(:protected_methods)
      when :private, :all
        list.concat bind_call(:private_methods)
      end
    end

    return list
  end

  # Is the object an instance of a given class?
  #
  # Returns [Boolean]
  def of?(a_class)
    bind_call(:instance_of?, a_class)
  end

  # Is the object an instance of a given class or subclass?
  #
  # Returns [Boolean]
  def is_a?(a_class)
    bind_call(:is_a?, a_class)
  end

  # Is the object an instance of a given class or subclass?
  #
  # Returns [Boolean]
  def kind_of?(a_class)
    bind_call(:kind_of?, a_class)
  end

  # Is an instaance variable defined?
  #
  # Returns [Boolean]
  def variable_defined?(name)
    name = atize(name)
    bind_call(:instance_variable_defined?, name)
  end

  # Get object's instance id.
  #
  # Returns [Integer]
  def id
    bind_call(:object_id)
  end

  # Fallback to get the real class of the Instance delegate itself.
  alias :object_class :class

  # Get object's instance id.
  #
  # Returns [Class]
  def class
    bind_call(:class)
  end

  # Send message to instance.
  def send(*a, &b)
    bind_call(:send, *a, &b)
  end

private

  def atize(name)
    name.to_s !~ /^@/ ? "@#{name}" : name
  end

  ##
  # ModuleExtensions provides some additional methods for Module and Class
  # objects.
  #
  # TODO: Are there any other module/class methods that need to be provided?
  #
  module ModuleExtensions
    # Store Object methods so they cannot be overriden by the delegate class.
    METHODS = {}

    def self.freeze_method(name)
      METHODS[name.to_sym] = Module.instance_method(name)
    end

    freeze_method :instance_method
    freeze_method :instance_methods
    freeze_method :public_instance_methods
    freeze_method :protected_instance_methods
    freeze_method :private_instance_methods

    # List of method definitions in a module or class.
    #
    # selection - Any of `:public`, `:protected` or `:private` which
    #             is used to select specific subsets of methods.
    #
    # Returns [Array<Symbol>]
    def method_definitions(*selection)
      list = []

      if selection.empty?
        list.concat bind_call(:instance_methods)
      end

      selection.each do |s|
        case s
        when :public, :all
          list.concat bind_call(:public_instance_methods)
        when :protected, :all
          list.concat bind_call(:protected_instance_methods)
        when :private, :all
          list.concat bind_call(:private_instance_methods)
        end
      end

      return list
    end

    # Shorter alias for #method_definitions.
    alias :definitions :method_definitions

    # Get a first-class method definition object.
    #
    # Returns an unbound method object. [UnboundMethod] 
    def method_definition(name)
      bind_call(:instance_method, name)
    end

    alias :definition :method_definition
  end

  # TODO: Are there any method we need specific to a Class vs a Module?
  #module ClassExtensions
  #
  #end

  private

  # Helper method for binding the method to the delegate object and calling it.
  def bind_call(method_name, *args, &block) 
    METHODS[mathod_name].bind(@delegate).call(&args, &block)
  end
end


class BasicObject
  # Returns an instance of Instance for `self`, which allows convenient
  # access to an object's internals.
  def instance
    ::Instance.instance(self)
  end
end

